### ISSUE TITLE
"Remote end closed socket abruptly" Node.js 12.16.3

### ISSUE THREAD

**Issue by ivanstanev** [tomas/needle#312]
Hey! The new patch release of Node.js (12.16.3) broke needle (we're using 2.4.0) and using needle with redirects now causes the following error:

```js
Error: Remote end closed socket abruptly.
    at TLSSocket.on_socket_end (/Users/ivan/local/node_modules/needle/lib/needle.js:473:17)
    at processTicksAndRejections (internal/process/task_queues.js:79:11)
```

You can reproduce it with the following piece of code:
```js
await needle('get',
  `https://github.com/kubernetes-sigs/kind/releases/download/v0.7.0/kind-linux-amd64`,
  null,
  { follow_max: 2 },
);
```

A similar issue was raised a while ago here: https://github.com/tomas/needle/issues/261

**Comment by tomas** [comment]
Darn. Do you know what did the Node.js release change regarding TLS sockets?

**Comment by gitphill** [comment]
https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V12.md#12.16.3, Updated OpenSSL to 1.1.1g?

**Comment by tomas** [comment]
Will take a look at this later today.

**Comment by ronag** [comment]
https://github.com/nodejs/node/issues/33254

**Comment by ronag** [comment]
[This condition looks wrong](https://github.com/tomas/needle/blob/master/lib/needle.js#L489).

It worked by luck before due to bug in Node that was fixed in https://github.com/nodejs/node/commit/ed21d32a7c37d030ce4668415832a89dc8ead34c#diff-e6ef024c3775d787c38487a6309e491dL634.

**Comment by tomas** [comment]
Good call @ronag. So we should remove that check altogether?

**Comment by ronag** [comment]
I mean I think it's a good idea. But should probably be implemented something along the lines of:

```js
const { finished } = require('stream-readable')

finished(this.socket, (err) => {
  if (err.code === 'ERR_STREAM_PREMATURE_CLOSE') {
      this.destroy();
      had_error(new Error('Remote end closed socket abruptly.'))
  }
})

```

**Comment by ronag** [comment]
> Good call @ronag. So we should remove that check altogether?

Yes. Use `response.on('aborted', handler)` instead.

**Comment by mhio** [comment]
Wow.. this has been driving me crazy for a week. 

For some reason the issue seems to be exacerbated by the builds running in a CI container. I couldn't reproduce the issue in a local docker build `FROM node:12.16.3` container, but via a `kaniko` or `buildah` container `node-pre-gyp` downloads using needle would fail every time.

**Comment by zbjornson** [comment]
Sorry to pester, but any chance this can be resolved soon? We're getting a lot of issues opened in node-canvas because our prebuilds are failing to download. Thank you!

**Comment by tomas** [comment]
@zbjornson Looking at it right now. Sorry for the delay.

**Comment by tomas** [comment]
Ok looks like I found a fix. Once I get the last three specs to pass on Node v12.16.3 I'll push a new version.

**Comment by marcosflick** [comment]
Any updates on this?

**Comment by tomas** [comment]
Yes! I'll push a new version in a while.

**Comment by tomas** [comment]
Ok we're good to go.

![image](https://user-images.githubusercontent.com/10574/82526690-95948580-9b02-11ea-8892-509bea0b49a4.png)

**Comment by tomas** [comment]
Just pushed 2.5.0 which fixes this issue.

Sorry everyone for the long delay -- I run an [ecommerce platform](https://wwwbootic.io) and these past weeks have been a bit crazy. 

PS: @ronag I tried the `aborted` callback or the `stream.finished` route but they didn't work, so I ended up adding an simple check that does the trick, while still handling the scenario tested in `test/long_string_spec.js`. Feel free to submit a PR if you find a better strategy though. :)
